src/ReasonReactRouter.re

```diff
@@ -67,50 +67,68 @@ let arrayToList = a => {
 /* sigh URLSearchParams doesn't work on IE11, edge16, etc. */
 /* actually you know what, not gonna provide search for now. It's a mess.
    We'll let users roll their own solution/data structure for now */
-let path = () =>
-  switch ([%external window]) {
-  | None => []
-  | Some((window: Dom.window)) =>
-    switch (window |> location |> pathname) {
-    | ""
-    | "/" => []
-    | raw =>
-      /* remove the preceeding /, which every pathname seems to have */
-      let raw = Js.String.sliceToEnd(~from=1, raw);
-      /* remove the trailing /, which some pathnames might have. Ugh */
-      let raw =
-        switch (Js.String.get(raw, Js.String.length(raw) - 1)) {
-        | "/" => Js.String.slice(~from=0, ~to_=-1, raw)
-        | _ => raw
-        };
-      raw |> Js.String.split("/") |> arrayToList;
-    }
+let hsMatcher = [%re "/(#[^?&]+)?([?&].*)?$/"];
+let pathMatcher = [%re "/(#|\\?).+$/"];
+let pathParse = str =>
+  switch (str) {
+  | ""
+  | "/" => []
+  | raw =>
+    /* remove the preceeding /, which every pathname seems to have */
+    let raw = Js.String.sliceToEnd(~from=1, raw);
+    /* remove the trailing /, which some pathnames might have. Ugh */
+    let raw =
+      switch (Js.String.get(raw, Js.String.length(raw) - 1)) {
+      | "/" => Js.String.slice(~from=0, ~to_=-1, raw)
+      | _ => raw
+      };
+    raw
+    |> Js.String.replaceByRe(pathMatcher, "")
+    |> Js.String.split("/")
+    |> Js.Array.filter(item => String.length(item) != 0)
+    |> arrayToList;
   };
-let hash = () =>
-  switch ([%external window]) {
-  | None => ""
-  | Some((window: Dom.window)) =>
-    switch (window |> location |> hash) {
-    | ""
-    | "#" => ""
-    | raw =>
-      /* remove the preceeding #, which every hash seems to have.
-         Why is this even included in location.hash?? */
-      raw |> Js.String.sliceToEnd(~from=1)
+
+let path = (~serverUrlString=None, ()) =>
+  switch (serverUrlString, [%external window]) {
+  | (None, None) => []
+  | (Some(serverUrlString), _) => pathParse(serverUrlString)
+  | (_, Some((window: Dom.window))) =>
+    pathParse(window |> location |> pathname)
+  };
+let hashParse = str =>
+  switch (str) {
+  | ""
+  | "#" => ""
+  | raw =>
+    switch (raw |> Js.String.splitByReAtMost(hsMatcher, ~limit=3)) {
+    | [|_, Some(hash), _|] => hash |> Js.String.sliceToEnd(~from=1)
+    | _ => ""
     }
   };
-let search = () =>
-  switch ([%external window]) {
-  | None => ""
-  | Some((window: Dom.window)) =>
-    switch (window |> location |> search) {
-    | ""
-    | "?" => ""
-    | raw =>
-      /* remove the preceeding ?, which every search seems to have. */
-      raw |> Js.String.sliceToEnd(~from=1)
+let hash = (~serverUrlString=None, ()) =>
```

Comment:
Hash isn't sent to the server so no point in adding the overhead for support here.

Reply:
Oh gotcha, less code to maintain :)